vimoutliner/scripts/votl_maketags.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH vo_maketags 1 "August 13, 2003"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
vo_maketags \- build TAGS for vim outliner
.SH SYNOPSIS
.B vo_maketags
.RI [ filename ]
.SH DESCRIPTION
This manual page documents briefly the
.B vo_maketags
Perl script.
.PP
.B vo_maketags
is a
.BR perl (1)
script that creates tag file which works with
.B vimoutliner
.RB ( vim (1)
plugin for editing outlines).
.SH OPTIONS
.B vo_maketags
script called just with filename of the processed outline, which
is a top of the hierarchy of linked outlines.
.SH SEE ALSO
.BR vim (1)
.ig
.br
The script is fully documented on
.UR http://www.vimoutliner.org
http://www.vimoutliner.org
.UE
..
.SH AUTHOR
This manual page was written by Matej Cepl <matej@ceplovi.cz>,
for the Debian GNU/Linux system (but may be used by others).


CHANGELOG
CHANGELOG

Version 0.3.7
Released 2014-04-05
	* Additions/Changes:
	- Added new hoisting plugin.
	- Added new tags plugin.
	- Add weighted checkboxes.
	- Now checkboxes calculations will ignore dash ([-]) marked items.
	- Added new math plugin.
	- Merged new clock plugin.
	- Added new format plugin.
	- Added new gtd plugin.
	- Added updated smart_paste plugin.
	- Added otl2latex script with files.
	- New file type naming convention: vo_base to votl.
	- Color fixes and updates.
	- Updated vimoutlinerrc with more options for new plugins.
	- Moved previous hoist plugin to prev directory,
	- Made new TODO list
	- Uploaded MediaWiki2otl script. 
	- Updated documentation.
	* Bug fixes:
	- Stop setting global options.
	- Fixed checkboxes calculation to make it faster and more consistent.
	- Strtime fix in vo_base.vim
	- Corrections to READNE.detailed, readme, INSTALL and cheatsheet files.
	- Corrections to install.sh.
	- Replaced len() so it works with multibyte chars.

Version 0.3.6
Released 2011-05-09
	* Additions/Changes:
	- Added updated comments and information to the user in the INSTALL
	  file.
	- Added comments and information to the user in install.sh to align
	  with the new version.
	- Added new light version of README and moved the more extensive file
	  to README.detailed
	- Several updates and additions to the README, README.detailed and
	  doc/vo_readme.txt file.
	* Bug fixes:
	- Source only the first vimoutlinerrc found.
	- Do not echo mappings.
	- Fixed typo and updated reference in install.sh.
	- Fixed several typos and updated reference in the
	  ftplugin/vo_base.vim
	- Fixed typos and removed outdated references in the vo_hoist.otl and
	  vo_hoist.vim files.
	- Fixed typos and removed outdated references in the vo_checkbox.otl
	  file.
	- Fixed two mappings for sourcing rc file and writing file in
	  ftplugin/vo_base.vim. 
	- Fixed typos in colors/vo_base.vim
	- Fixed errors when g:vo_modules_load is not defined, also updated the
	  plugin loading process to use lists.

Version 0.3.5
Released 2011-03-20
	* Additions/Changes:
	- Moved all files inside $HOME/.vim.
	- Added a cheat sheet to the documentation.
	- Added smart-paste plugin.
	- Added and updated post-processor scripts:
		· fs2otl
		· Node.pm
		· otl2aft.awk
		· otl2docbook.pl
		· otl2html.py
		· otl2lyx.awk
		· otl2ooimpress.py
		· otl2table.py
		· otl2tags.py
		· otlgrep.py
		· otlhead.sh
		· otlsplit.py
		· otltail.sh
		· otl_handler
		· outline_calendar
		· outline_freemind
		· vo_maketags.pl
	- Added syntax support for _ilink_.
	- Hoist plugin disabled by default.
	- Layout of vo_readme.txt made more similar to Vim help.
	* Bug fixes:
	- ,,S did not work. Fixed

Version 0.3.4
Released
	* Additions/changes:
	- added ,,cp as an alias to ,,c%
	- Color schemes have been added
	- Checkboxes and hoisting default to 'on'
	- Modified ,,cb and ,,c% (and ,,cp) to work only on headings
	- Fixed the ,,cz command to make the correct call
	- Added descriptions of VO objects to help (headings, text, tables,
	  etc.)
	* Bug fixes:
	- W18 errors have been fixed

=============================================================================
RCS file: /root/VimOutliner/VimOutliner/INSTALL,v
Working file: INSTALL
head: 1.2
branch:
locks: strict
access list:
symbolic names:
keyword substitution: kv
total revisions: 2;	selected revisions: 2
description:
----------------------------
revision 1.2
date: 2004-11-27 19:30:43 +0000;  author: noel;  state: Exp;  lines: +2 -2

Minor syntax and installation adjustments.
Change to vo_checkbox to fix a typo.
----------------------------
revision 1.1
date: 2004-02-17 22:00:15 +0000;  author: noel;  state: Exp;
These files are only in cvs for backup purposes. They should be extracted
from vo_readme.txt and never edited themselves:
  INSTALL
  LICENSE
  VERSION
=============================================================================

RCS file: /root/VimOutliner/VimOutliner/Attic/INSTALL.TXT,v
Working file: INSTALL.TXT
head: 1.3
branch:
locks: strict
access list:
symbolic names:
keyword substitution: kv
total revisions: 3;	selected revisions: 3
description:
Quick installation instructions for developers.
----------------------------
revision 1.3
date: 2003-06-20 14:13:15 +0000;  author: noel;  state: dead;  lines: +0 -0
Final documentation tweaks for the final 0.3.0 release.
----------------------------
revision 1.2
date: 2003-06-18 13:50:55 +0000;  author: noel;  state: Exp;  lines: +1 -1
Modified the installation title.
----------------------------
revision 1.1
date: 2003-03-20 22:56:49 +0000;  author: noel;  state: Exp;
Initial revision
=============================================================================

RCS file: /root/VimOutliner/VimOutliner/LICENSE,v
Working file: LICENSE
head: 1.2
branch:
locks: strict
access list:
symbolic names:
keyword substitution: kv
total revisions: 2;	selected revisions: 2
description:
----------------------------
revision 1.2
date: 2004-11-27 19:30:43 +0000;  author: noel;  state: Exp;  lines: +4 -2

Minor syntax and installation adjustments.
Change to vo_checkbox to fix a typo.
----------------------------
revision 1.1
date: 2004-02-17 22:00:15 +0000;  author: noel;  state: Exp;
These files are only in cvs for backup purposes. They should be extracted
from vo_readme.txt and never edited themselves:
  INSTALL
  LICENSE
  VERSION
=============================================================================

RCS file: /root/VimOutliner/VimOutliner/VERSION,v
Working file: VERSION
head: 1.4
branch:
locks: strict
access list:
symbolic names:
keyword substitution: kv
total revisions: 4;	selected revisions: 4
description:
----------------------------
revision 1.4
date: 2005-06-07 17:48:50 +0000;  author: noel;  state: Exp;  lines: +4 -0
Updated to include help on VO objects.
----------------------------
revision 1.3
date: 2005-06-07 17:41:42 +0000;  author: noel;  state: Exp;  lines: +10 -10
Updated version.
----------------------------
revision 1.2

vimoutliner/scripts/votl_maketags.pl
#!/usr/bin/perl -w
# #######################################################################
# vo_maketags.pl: Vim outline tagging system, main program, version 0.3.5
#   Copyright (C) 2001-2003, 2011 by Steve Litt (slitt@troubleshooters.com)
#
#   This program is free software; you can redistribute it and/or modify
#   it under the terms of the GNU General Public License as published by
#   the Free Software Foundation; either version 2 of the License, or
#   (at your option) any later version.
#
#   This program is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#   GNU General Public License for more details.
#
#   You should have received a copy of the GNU General Public License
#   along with this program; if not, see <http://www.gnu.org/licenses/>.
#
# Steve Litt, slitt@troubleshooters.com, http://www.troubleshooters.com
# #######################################################################

# #######################################################################
# #######################################################################
# #######################################################################
# HISTORY
# V0.1.0 Pre-alpha
#     Starting at a "top level" indent-defined Vim outline, this
#     program finds all "tags" defined as headlines starting with
#     _tag_, and containing a subheadline containing the file
#     to which the tag should jump. This program creates a tags
#     file.
#Steve Litt, 5/28/2001
#End of version 0.1.0
#
# V0.1.1 Pre-alpha
#     Bug fixes, including ../ resolution
#
#Steve Litt, 5/28/2001
#End of version 0.1.1
#
#
# V0.1.2 Pre-alpha
#	More bug fixes, and facility to create a new outline
#	file from a tag whose corresponding file doesn't yet
#	exist.  
#Steve Litt, 5/30/2001
#End of version 0.1.2
#
# V0.1.3 Pre-alpha
#	More bug fixes. This was the first version released
#	file from a tag whose corresponding file doesn't yet
#	exist.  
#Steve Litt, 6/01/2001
#End of version 0.1.3
#
# V0.2.0 Pre-alpha
#Steve Litt, 12/03/2002
#	This file unchanged. The overall Vimoutliner version
#	0.2.0 has extensive improvements, including intuitive
#	collapse/expand.
#End of version 0.2.0
#END OF HISTORY
#
#
# V0.1.2 Pre-alpha
#	More bug fixes, and facility to create a new outline
#	file from a tag whose corresponding file doesn't yet
#	exist.  
#Steve Litt, 5/30/2001
#End of version 0.1.2
# V0.3.5 release 20110303
#       Changed vo_tags.tag directory from
#       $HOME/.vimoutliner/
#       to
#       $HOME/.vim/vimoutliner/
#Steve Litt, 3/3/2011
#End of version 0.3.5 release 20110303
#END OF HISTORY
#
# #######################################################################


use strict;
use vars qw($TAGFILENAME);
use Cwd;

$TAGFILENAME = $ENV{"HOME"} . "/.vim/vimoutliner/vo_tags.tag";
##### OLD LOCATION BELOW, REMOVE IN 0.3.6, COMMENT FOR NOW
#$TAGFILENAME = $ENV{"HOME"} . "/.vimoutliner/vo_tags.tag";

sub process1Outline($$); #Early prototype the recursive routine
sub makeDirectory($);    #Early prototype the recursive routine

sub makeTagFileStartingAt($)
	{
	unless(@ARGV == 1)
		{
		usage();
		die;
		}
	my($absoluteFileName) =  deriveAbsoluteFileName(Cwd::cwd(), $_[0]);
	
	my(%processedFiles) = ();
	recordFileAsProcessed($absoluteFileName,\%processedFiles);
	unlink $TAGFILENAME;
	process1Outline($absoluteFileName, \%processedFiles);
	sortAndDeleteDupsFromTagFile();
	}

sub sortAndDeleteDupsFromTagFile()
	{
	my($TEMPTAGFILENAME) = "$ENV{'HOME'}/temptagfile.tag";
	system("sort $TAGFILENAME | uniq > $TEMPTAGFILENAME"); 
	system("rm $TAGFILENAME");
	system("mv $TEMPTAGFILENAME $TAGFILENAME");
	}


sub process1Outline($$)
	{
	my($fileName) = $_[0];
	my($processedFilesHashRef) = $_[1];
	
	unless(fileExists($fileName))
		{
		makeDirectory($fileName);
		makeEmptyFile($fileName);
		}

	print "Begin processing file $fileName.\n";

	my($baseDirectory) = getBaseDirectory($fileName);
	my(%tags) = getTagsFromFile($fileName);
	my(@tagKeys) = keys(%tags);
	my($tagKey);
	foreach $tagKey (@tagKeys)
		{
		my($absoluteFileName);
		if(isAbsoluteFilePath($tags{$tagKey}))
			{
			$absoluteFileName = $tags{$tagKey};
			}
		else
			{
			$absoluteFileName =
			 deriveAbsoluteFileName($baseDirectory, $tags{$tagKey});
			}
		appendTagToTagFile($tagKey,$absoluteFileName);
 		if(notProcessedYet($absoluteFileName, $processedFilesHashRef))
 			{
			recordFileAsProcessed($absoluteFileName,$processedFilesHashRef);
 			process1Outline($absoluteFileName, $processedFilesHashRef);
 			}
		}
	}

sub appendTagToTagFile($$)
	{
	open(TAGFILE, ">>$TAGFILENAME");
	print TAGFILE "$_[0]	$_[1]	:1\n";
	close(TAGFILE);
	}


sub makeEmptyFile($)
	{
	open(OUTLINEFILE, ">" . $_[0]);
	close(OUTLINEFILE);
	}


sub makeDirectory($)
	{
	my($completeFileName) = $_[0];
	my($directoryName) = ($completeFileName =~ m/^(.*?)[^\/]*$/);
	unless($directoryName eq "")
		{
		my($temp) = ($directoryName =~ m/^(.*).$/);
		makeDirectory($temp);
		print "Creating $directoryName...";
		if(mkdir $directoryName)
			{
			print " succeeded.\n";
			}
		else
			{
			print " no action: $!.\n";
			}
		}
	}

sub fileExists($)
	{
	my($outlineFileName) = $_[0];
	my($success) = open(OUTLINEFILE, "<" . $outlineFileName);
	if($success)
		{
		close(OUTLINEFILE);
		return(1);
		}
	else
		{
		return(0);
		}
	}

sub getTagsFromFile($)
	{
	my($outlineFileName) = $_[0];
	my(%tags);
	my($tagString) = "";
	my($success) = open(OUTLINEFILE, "<" . $outlineFileName);
	unless($success)
		{
		print "Failed to open $outlineFileName\n";
		return(());
		}
	while(<OUTLINEFILE>)
		{
		my($line) = $_;
		chomp($line);
		if($line =~ m/^\s*(_tag_\S+)/)
			{
			$tagString = $1;
			}
		elsif($tagString ne "")
			{
			$line =~ m/^\s*(\S+)/;
			my($filename) = $1;
			$tags{$tagString} =
			  deriveAbsoluteFileName(getBaseDirectory($_[0]), $1);
			$tagString = "";
			}
		}	
	return(%tags);
	}

sub recordFileAsProcessed($$)
	{
	my($absoluteFileName) = $_[0];
	my($processedFilesHashRef) = $_[1];
	${$processedFilesHashRef}{$absoluteFileName} = "1";
	}

sub notProcessedYet($$)
	{
        my($absoluteFileName) = $_[0];
	my(%processedFiles) = %{$_[1]};
	if(defined($processedFiles{$absoluteFileName}))
		{
		return(0);
		}
	else
		{
		return(1);
		}
	}

sub dia($)
	{
	print "dia " . $_[0] . "\n";
	}


sub isAbsoluteFilePath($)
	{
	if($_[0] =~ m/^\//)
		{
		return 1;
		}
	else
		{
		return 0;
		}
	}

sub getFileNameOnly($)
	{ 
	my($fileString); 
	if ($_[0] =~ m/.+\/(.*)$/)
		{
		$fileString= $1
		}
	else
		{
		$fileString = $_[0];
		}

	return $fileString;
	}

sub getBaseDirectory($)
	{ 
	my($dirString) = ($_[0] =~ m/(.+\/).*$/);
	return $dirString;
	}

sub deriveAbsoluteFileName($$)
	{
	my($absoluteFileName);
	my($baseDirectory) = $_[0];
	my($passedFileName) = $_[1];
	unless($baseDirectory =~ m/\/$/)
		{
		$baseDirectory= $baseDirectory . "/"; 
		}
	if($passedFileName =~ m/^\//)
		{
		$absoluteFileName = $passedFileName;
		}
	else
		{
		$absoluteFileName = $baseDirectory . $passedFileName;
		}

	$absoluteFileName =~ s/\/\.\//\//g;  #remove all "./";
	deleteDoubleDots($absoluteFileName);

	return($absoluteFileName);
	}

sub deleteDoubleDots($)
	{
	while($_[0] =~ m/\.\./)
		{
		$_[0] =~ s/\/[^\/]*\/\.\.//;
		}
	}

sub usage()
	{
	print "\nUsage is:\n";
	print "otltags topLevelOutlineFileName\n\n";
	}
	

makeTagFileStartingAt($ARGV[0])

